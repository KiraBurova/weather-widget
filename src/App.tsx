import React, { useState, useEffect } from 'react';

import { getWeatherIconByIconName, getWeatherReportByLocation } from './api';

import MainWidget from './components/Widget';

import { WeatherReportType, WidgetDataType } from './types';

import styles from './App.module.scss';

const App = () => {
  const [error, setError] = useState('');
  const [weatherReports, setWeatherReports] = useState<Array<WidgetDataType>>([]);

  useEffect(() => {
    const handleCheckIfGeoLocationIsAvailable = () => {
      if ('geolocation' in navigator) {
        handleGetCurrentLocation();
      } else {
        setError('Geolocation is not available');
      }
    };
    handleCheckIfGeoLocationIsAvailable();
  }, []);

  const handleTransformLocationData = (weatherData: WeatherReportType) => {
    return {
      cityName: weatherData.name,
      countryName: weatherData.sys.country,
      weatherDescription: weatherData.weather[0].description,
      feelsLike: weatherData.main.feels_like,
      weatherIcon: getWeatherIconByIconName(weatherData.weather[0].icon),
      temperature: weatherData.main.temp,
      wind: weatherData.wind,
      humidity: weatherData.main.humidity,
      visibility: weatherData.visibility,
      pressure: weatherData.main.pressure,
      clouds: weatherData.clouds.all,
    };
  };

  const handleGetCurrentLocation = () => {
    navigator.geolocation.getCurrentPosition((position) => {
      const { latitude, longitude } = position.coords;

      getWeatherReportByLocation(latitude, longitude)
        .then((response) => response.json())
        .then((weatherData: WeatherReportType) => {
          const weatherReport: WidgetDataType = handleTransformLocationData(weatherData);
          console.log(weatherData);
          setWeatherReports([...weatherReports, weatherReport]);
        })
        .catch((error: TypeError) => setError(error.message));
    });
  };

  return (
    <div className={styles.app}>
      {!!weatherReports.length && weatherReports.map((weatherReport) => <MainWidget weatherReport={weatherReport} />)}
      <span className={styles.error}>{error}</span>
    </div>
  );
};

export default App;
