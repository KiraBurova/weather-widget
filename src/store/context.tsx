import { useLocalObservable } from 'mobx-react-lite';
import React from 'react';

import { createStore, TStore } from './store';

const StoreContext = React.createContext<TStore | null>(null);

type StoreProviderProps = {
  children: React.ReactNode;
};

export const StoreProvider = ({ children }: StoreProviderProps) => {
  const store = useLocalObservable(createStore);
  return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};

export const useStore = () => {
  const store = React.useContext(StoreContext);
  if (!store) {
    throw new Error('useStore must be used within a StoreProvider.');
  }
  return store;
};
