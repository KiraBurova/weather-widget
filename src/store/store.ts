import { getWeatherReportByCityName, getWeatherReportByLocation, getWeatherIconByIconName } from '../api';

import { TLocation, WeatherReportType } from '../types';

export function createStore() {
  return {
    location: '',
    locations: [] as TLocation[],
    weatherReports: [] as WeatherReportType[],
    fetchWeatherReportByLocation(latitude: number, longitude: number) {
      getWeatherReportByLocation(latitude, longitude)
        .then((weatherData: any) => {
          console.log(weatherData);
          const w = this.transformLocationData(weatherData.data);
          this.setWeatherReports(w);
        })
        .catch((error: Error) => console.log(error));
    },
    fetchWeatherReportByCityName(location: TLocation) {
      getWeatherReportByCityName(location.name)
        .then((weatherData) => {
          this.setLocations(location);
          const w = this.transformLocationData(weatherData.data);
          this.setWeatherReports(w);
        })
        .catch((error: Error) => console.log(error));
    },
    setLocations(location: TLocation) {
      this.locations.push(location);
    },
    deleteLocation(id: number) {
      this.locations = this.locations.filter((location) => location.id !== id);
    },
    setWeatherReports(weatherReport: any) {
      this.weatherReports.push(weatherReport);
    },
    transformLocationData(weatherData: WeatherReportType) {
      return {
        cityName: weatherData.name,
        countryName: weatherData.sys.country,
        weatherDescription: weatherData.weather[0].description,
        feelsLike: weatherData.main.feels_like,
        weatherIcon: getWeatherIconByIconName(weatherData.weather[0].icon),
        temperature: weatherData.main.temp,
        wind: weatherData.wind,
        humidity: weatherData.main.humidity,
        visibility: weatherData.visibility,
        pressure: weatherData.main.pressure,
        clouds: weatherData.clouds.all,
      };
    },
  };
}
export type TStore = ReturnType<typeof createStore>;
