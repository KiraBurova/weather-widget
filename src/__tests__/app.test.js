import React from 'react';

import { render, fireEvent, waitFor } from '@testing-library/react';

import { StoreProvider } from '../store/context';

import App from '../App';

describe('<App />', () => {
  it('show error when no geolocation is available', () => {
    const { container, getByTestId } = render(
      <StoreProvider>
        <App></App>
      </StoreProvider>,
    );
    const error = getByTestId('error');

    expect(error).toBeTruthy();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('show settings', async () => {
    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementationOnce((success) =>
        Promise.resolve(
          success({
            coords: {
              latitude: 51.1,
              longitude: 45.3,
            },
          }),
        ),
      ),
    };
    global.navigator.geolocation = mockGeolocation;

    const { container, getByRole, getByText } = render(
      <StoreProvider>
        <App></App>
      </StoreProvider>,
    );

    await waitFor(() => {
      const button = getByRole('button');

      fireEvent.click(button);

      expect(getByText('Settings')).toBeTruthy();
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('show error when geolocation is not given', async () => {
    const mockGeolocation = {
      getCurrentPosition: jest.fn().mockImplementationOnce((success, error) =>
        Promise.resolve(
          error({
            code: 1,
            message: 'GeoLocation Error',
          }),
        ),
      ),
    };

    global.navigator.geolocation = mockGeolocation;

    const { container, getByTestId } = render(
      <StoreProvider>
        <App></App>
      </StoreProvider>,
    );

    await waitFor(() => {
      expect(getByTestId('error')).toBeTruthy();
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
