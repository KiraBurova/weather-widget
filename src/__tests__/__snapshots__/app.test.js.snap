// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> show error when geolocation is not given 1`] = `
<div
  class="app"
>
  <div
    class="container"
  >
    <span
      class="error"
      data-testid="error"
    >
      GeoLocation Error
    </span>
  </div>
</div>
`;

exports[`<App /> show error when no geolocation is available 1`] = `
<div
  class="app"
>
  <div
    class="container"
  >
    <span
      class="error"
      data-testid="error"
    >
      Geolocation is not available
    </span>
  </div>
</div>
`;

exports[`<App /> show settings 1`] = `
<div
  class="app"
>
  <div
    class="container"
  >
    <div
      class="settings"
    >
      <div
        class="settingsTop"
      >
        <span>
          Settings
        </span>
        <div>
          <svg
            class="closeIcon"
            focusable="false"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style="transform: rotate(360deg);"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5L12 10.59L6.41 5L5 6.41L10.59 12L5 17.59L6.41 19L12 13.41L17.59 19L19 17.59L13.41 12z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="addedLocations"
        data-rbd-droppable-context-id="0"
        data-rbd-droppable-id="droppable"
      >
        <div
          data-rbd-draggable-context-id="0"
          data-rbd-draggable-id="0.3393428674592154"
          style="margin-bottom: 5px;"
        >
          <div
            class="item"
          >
            <div
              aria-describedby="rbd-hidden-text-0-hidden-text-0"
              data-rbd-drag-handle-context-id="0"
              data-rbd-drag-handle-draggable-id="0.3393428674592154"
              draggable="false"
              role="button"
              tabindex="0"
            >
              <svg
                class="dragIcon"
                focusable="false"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style="transform: rotate(360deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
                  fill="currentColor"
                />
              </svg>
            </div>
            Kamenskiy
            <span
              class="deleteIcon"
              role="button"
            >
              <svg
                focusable="false"
                height="1em"
                preserveAspectRatio="xMidYMid meet"
                style="transform: rotate(360deg);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM8 9h8v10H8V9zm7.5-5l-1-1h-5l-1 1H5v2h14V4h-3.5z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div>
        <span
          class="formLabel"
        >
          Add location:
        </span>
        <form
          class="form"
          data-testid="form"
        >
          <input
            class="locationInput"
            value=""
          />
          <button
            class="addButton"
            disabled=""
            type="submit"
          >
            <svg
              focusable="false"
              height="1em"
              preserveAspectRatio="xMidYMid meet"
              style="transform: rotate(360deg);"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11 9l1.42 1.42L8.83 14H18V4h2v12H8.83l3.59 3.58L11 21l-6-6l6-6z"
                fill="currentColor"
              />
            </svg>
          </button>
        </form>
      </div>
    </div>
    
  </div>
</div>
`;
